src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/EnvelopeEventProcessor.java

```diff
@@ -70,18 +83,82 @@ private void process(IMessage message) {
             );
 
             eventPublisher.publish(envelope);
-
             log.info("Processed message with ID {}", message.getMessageId());
+            return new MessageProcessingResult(SUCCESS);
+        } catch (InvalidMessageException ex) {
+            log.error("Rejected message with ID {}, because it's invalid", message.getMessageId(), ex);
+            return new MessageProcessingResult(UNRECOVERABLE_FAILURE, ex);
         } catch (Exception ex) {
-            throw new MessageProcessingException(
-                "Failed to process message with ID " + message.getMessageId(),
+            log.error("Failed to process message with ID {}", message.getMessageId(), ex);
+            return new MessageProcessingResult(POTENTIALLY_RECOVERABLE_FAILURE);
+        }
+    }
+
+    private void tryFinaliseProcessedMessage(IMessage message, MessageProcessingResult processingResult) {
+        try {
+            finaliseProcessedMessage(message, processingResult);
+        } catch (Exception ex) {
+            log.error(
+                "Failed to manage processed message with ID {}. Processing result: {}",
+                message.getMessageId(),
+                processingResult,
                 ex
             );
         }
     }
 
+    private void finaliseProcessedMessage(
+        IMessage message,
+        MessageProcessingResult processingResult
+    ) throws InterruptedException, ServiceBusException {
+
+        switch (processingResult.resultType) {
+            case SUCCESS:
+                messageOperations.complete(message.getLockToken());
+                log.info("Message with ID {} has been completed", message.getMessageId());
+                break;
+            case UNRECOVERABLE_FAILURE:
+                messageOperations.deadLetter(
+                    message.getLockToken(),
+                    "Message processing error",
+                    processingResult.exception.getMessage()
+                );
+
+                log.info("Message with ID {} has been dead-lettered", message.getMessageId());
+                break;
+            case POTENTIALLY_RECOVERABLE_FAILURE:
+                // do nothing - let the message lock expire
+                log.info("Allowing message with ID {} to return to queue", message.getMessageId());
+                break;
+            default:
+                throw new MessageProcessingException(
+                    "Unknown message processing result type: " + processingResult.resultType
+                );
+        }
+    }
+
     @Override
     public void notifyException(Throwable exception, ExceptionPhase phase) {
         log.error("Error while handling message at stage: " + phase, exception);
     }
+
+    class MessageProcessingResult {
+        public final MessageProcessingResultType resultType;
+        public final Exception exception;
+
+        public MessageProcessingResult(MessageProcessingResultType resultType) {
```

Comment:
should it be legal to create a result of type `UNRECOVERABLE_FAILURE` without providing an exception?

Reply:
no, why?